File: AFD/automata.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Estructura para representar un estado del AFD
typedef struct {
    char tipo;       // '>' = inicial, '+' = aceptación, '-' = normal
    char nombre;     // nombre del estado (ej. 'A', 'B', 'C')
    char destino0;   // estado destino con símbolo '0'
    char destino1;   // estado destino con símbolo '1'
} Estado;

// Prototipos de funciones
int contarEstados(FILE* archivo);
void mostrarAlfabeto(FILE* archivo);
Estado* cargarEstados(FILE* archivo, int numEstados);
void mostrarTabla(Estado* estados, int numEstados);
void verificarCadenas(Estado* estados, int numEstados, FILE* archivoCadenas);
int buscarIndiceEstado(char nombreEstado, Estado* estados, int numEstados);

int main() {
    FILE *archivoConfig, *archivoCadenas;
    int numEstados;
    Estado* estados;

    // Abrir archivo de configuración
    archivoConfig = fopen("config.txt", "r");
    if(archivoConfig == NULL) {
        perror("Error al abrir config.txt");
        exit(1);
    }

    // Abrir archivo con cadenas de prueba
    archivoCadenas = fopen("prueba.txt", "r");
    if(archivoCadenas == NULL) {
        perror("Error al abrir prueba.txt");
        exit(1);
    }

    // Procesar configuración del AFD
    numEstados = contarEstados(archivoConfig);
    mostrarAlfabeto(archivoConfig);
    estados = cargarEstados(archivoConfig, numEstados);

    // Mostrar tabla de transiciones
    mostrarTabla(estados, numEstados);

    // Verificar las cadenas
    verificarCadenas(estados, numEstados, archivoCadenas);

    // Liberar memoria y cerrar archivos
    free(estados);
    fclose(archivoConfig);
    fclose(archivoCadenas);

    return 0;
}

// Cuenta cuántos estados hay antes del separador "-"
int contarEstados(FILE* archivo) {
    char buffer[8];
    int contador = -1;
    do {
        fgets(buffer, 8, archivo);
        contador++;
    } while(buffer[0] != '-');
    fseek(archivo, 0, SEEK_SET); // rebobinar
    return contador;
}

// Muestra el alfabeto definido en el archivo después de "-"
void mostrarAlfabeto(FILE* archivo) {
    char linea[10], simbolo;
    do {
        fgets(linea, 8, archivo);
    } while(linea[0] != '-');

    printf("Alfabeto: ");
    do {
        simbolo = fgetc(archivo);
        printf("%c", simbolo);
    } while(simbolo != '\n');
    printf("\n");
    fseek(archivo, 0, SEEK_SET); // rebobinar
}

// Carga los estados y transiciones en memoria
Estado* cargarEstados(FILE* archivo, int numEstados) {
    Estado* estados = (Estado*) calloc(numEstados, sizeof(Estado));
    char buffer[8];
    for(int i=0; i<numEstados; i++) {
        fgets(buffer, 8, archivo);
        estados[i].tipo     = buffer[0];
        estados[i].nombre   = buffer[1];
        estados[i].destino0 = buffer[2];
        estados[i].destino1 = buffer[3];
    }
    return estados;
}

// Muestra la tabla de transiciones
void mostrarTabla(Estado* estados, int numEstados) {
    printf("\nTabla de transiciones:\n");
    printf("Tipo Estado  0   1\n");
    for(int i=0; i<numEstados; i++) {
        printf(" [%c]   %c    %c   %c\n", 
            estados[i].tipo, 
            estados[i].nombre, 
            estados[i].destino0, 
            estados[i].destino1);
    }
    printf("\n");
}

// Busca el índice de un estado por nombre
int buscarIndiceEstado(char nombreEstado, Estado* estados, int numEstados) {
    for(int i=0; i<numEstados; i++) {
        if(estados[i].nombre == nombreEstado) {
            return i;
        }
    }
    return -1; // no encontrado
}

// Verifica cada cadena en el archivo
void verificarCadenas(Estado* estados, int numEstados, FILE* archivoCadenas) {
    char cadena[100];

    while(fgets(cadena, 100, archivoCadenas) != NULL) {
        cadena[strcspn(cadena, "\n")] = 0; // quitar salto de línea

        char estadoActual = '\0';
        // Buscar estado inicial
        for(int i=0; i<numEstados; i++) {
            if(estados[i].tipo == '>' || estados[i].tipo == '+') {
                estadoActual = estados[i].nombre;
                break;
            }
        }

        // Caso de cadena vacía (E)
        if(cadena[0] == 'E') {
            int idx = buscarIndiceEstado(estadoActual, estados, numEstados);
            if(idx != -1 && estados[idx].tipo == '+') {
                printf("Cadena \"%s\" aceptada (E)\n", cadena);
            } else {
                printf("Cadena \"%s\" no aceptada\n", cadena);
            }
            continue;
        }

        // Recorrer cada símbolo de la cadena
        for(int i=0; cadena[i] != '\0'; i++) {
            int idx = buscarIndiceEstado(estadoActual, estados, numEstados);
            if(idx == -1) break;

            if(cadena[i] == '0') {
                estadoActual = estados[idx].destino0;
            } else if(cadena[i] == '1') {
                estadoActual = estados[idx].destino1;
            }
        }

        // Verificar aceptación
        int idxFinal = buscarIndiceEstado(estadoActual, estados, numEstados);
        if(idxFinal != -1 && estados[idxFinal].tipo == '+') {
            printf("Cadena \"%s\" aceptada\n", cadena);
        } else {
            printf("Cadena \"%s\" no aceptada\n", cadena);
        }
    }
}

File: AFD/automata.py
class Estado:
    def __init__(self, tipo='', nombre='', destino0='', destino1=''):
        self.tipo = tipo
        self.nombre = nombre
        self.destino0 = destino0
        self.destino1 = destino1

def contar_estados(archivo_config):
    with open(archivo_config, 'r', encoding='utf-8') as archivo:
        contador = -1
        for linea in archivo:
            contador += 1
            if linea.strip().startswith('-'):
                break
    return contador

def mostrar_alfabeto(archivo_config):
    with open(archivo_config, 'r', encoding='utf-8') as archivo:
        # Buscar la línea que empieza con '-'
        for linea in archivo:
            if linea.strip().startswith('-'):
                break
        
        # La siguiente línea contiene el alfabeto
        alfabeto = archivo.readline().strip()
        print(f"Alfabeto: {alfabeto}")

def cargar_estados(archivo_config, num_estados):
    estados = []
    with open(archivo_config, 'r', encoding='utf-8') as archivo:
        for i in range(num_estados):
            linea = archivo.readline().rstrip('\n')  # Solo quitar salto de línea
            if len(linea) >= 4:
                tipo = linea[0] if linea[0] != ' ' else '-'  # Si es espacio, es estado normal
                nombre = linea[1]
                destino0 = linea[2]
                destino1 = linea[3]
                estados.append(Estado(tipo, nombre, destino0, destino1))
    return estados

def mostrar_tabla(estados):
    print("\nTabla de transiciones:")
    print("Tipo Estado  0   1")
    for estado in estados:
        print(f" [{estado.tipo}]   {estado.nombre}    {estado.destino0}   {estado.destino1}")
    print()

def buscar_indice_estado(nombre_estado, estados):
    for i, estado in enumerate(estados):
        if estado.nombre == nombre_estado:
            return i
    return -1

def buscar_estado_inicial(estados):
    for estado in estados:
        if estado.tipo == '>':
            return estado.nombre
    # Si no hay estado marcado como inicial, buscar uno que sea inicial Y de aceptación
    for estado in estados:
        if estado.tipo == '+':
            return estado.nombre
    return None

def verificar_cadenas(estados, archivo_cadenas):
    try:
        with open(archivo_cadenas, 'r', encoding='utf-8') as archivo:
            for linea in archivo:
                cadena = linea.strip()
                if not cadena:  # Saltar líneas vacías
                    continue
                
                # Buscar estado inicial
                estado_actual = buscar_estado_inicial(estados)
                
                if estado_actual is None:
                    print(f"Error: No se encontró estado inicial")
                    continue
                
                # Caso de cadena vacía (E)
                if cadena == 'E' or cadena == '':
                    idx = buscar_indice_estado(estado_actual, estados)
                    if idx != -1 and estados[idx].tipo == '+':
                        print(f'Cadena "{cadena}" aceptada (E)')
                    else:
                        print(f'Cadena "{cadena}" no aceptada')
                    continue
                
                # Procesar cada símbolo de la cadena
                for simbolo in cadena:
                    idx = buscar_indice_estado(estado_actual, estados)
                    if idx == -1:
                        break
                    
                    if simbolo == '0':
                        estado_actual = estados[idx].destino0
                    elif simbolo == '1':
                        estado_actual = estados[idx].destino1
                    else:
                        print(f'Error: Símbolo "{simbolo}" no reconocido en la cadena "{cadena}"')
                        break
                
                # Verificar si el estado final es de aceptación
                idx_final = buscar_indice_estado(estado_actual, estados)
                if idx_final != -1 and estados[idx_final].tipo == '+':
                    print(f'Cadena "{cadena}" aceptada')
                else:
                    print(f'Cadena "{cadena}" no aceptada')
    
    except FileNotFoundError:
        print(f"Error: No se pudo abrir el archivo {archivo_cadenas}")

def main():
    archivo_config = "Conf.txt"
    archivo_cadenas = "Cadenas.txt"
    
    try:
        # Procesar configuración del AFD
        num_estados = contar_estados(archivo_config)
        mostrar_alfabeto(archivo_config)
        estados = cargar_estados(archivo_config, num_estados)
        
        mostrar_tabla(estados)
        
        verificar_cadenas(estados, archivo_cadenas)
        
    except FileNotFoundError as e:
        print(f"Error al abrir archivo: {e}")
    except Exception as e:
        print(f"Error inesperado: {e}")

if __name__ == "__main__":
    main()
